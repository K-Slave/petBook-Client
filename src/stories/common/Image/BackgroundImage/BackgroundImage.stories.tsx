import { Meta, StoryObj } from "@storybook/react";
import BackgroundImage from "./BackgroundImage";

const meta: Meta<typeof BackgroundImage> = {
  title: "common/Image/BackgroundImage",
  component: BackgroundImage,
  tags: ["autodocs"],
  parameters: {
    layout: "fullscreen",
  },
};

type Story = StoryObj<typeof BackgroundImage>;

const originUrl =
  process.env.NODE_ENV === "production"
    ? "https://petbook.site"
    : "http://localhost:3000";

export const BgImgSpan: Story = {
  args: {
    url: `${originUrl}/icons/icons8-backjo-96.png`,
    width: "3.125rem",
    height: "3.125rem",
    bgPosition: "center",
    tag: "span",
  },
};

export const BgImgA: Story = {
  args: {
    url: `${originUrl}/icons/icons8-gosum-96.png`,
    width: "3.125rem",
    height: "3.125rem",
    bgPosition: "center",
    tag: "a",
    anchorAttr: {
      href: "https://icons8.com/icon/aKEeT9FEK2cQ/%EA%B3%A0%EC%8A%B4%EB%8F%84%EC%B9%98",
      target: "_blank",
      rel: "noopener",
    },
  },
};

export const BgImgDiv: Story = {
  args: {
    url: `${originUrl}/Landing__Background.png`,
    width: "18.75rem",
    height: "18.75rem",
    bgPosition: "center",
    tag: "div",
  },
};

export default meta;
