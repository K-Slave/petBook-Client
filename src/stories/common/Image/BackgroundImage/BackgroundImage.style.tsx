import styled, { CSSProperties, css } from "styled-components";

const BackgroundImageCommon = css`
  position: relative;

  image-rendering: -webkit-optimize-contrast;
  transform: translateZ(0);
  backface-visibility: hidden;
`;

export const BackgroundImageAttr = css`
  ${BackgroundImageCommon}
  background-size: contain;
  background-repeat: no-repeat;
  background-position: center;
`;

export const BackgroundImageSpan = styled.span<{
  url: string;
  width: string;
  height: string;
  display?: CSSProperties["display"];
  bgPosition?: CSSProperties["backgroundPosition"];
}>`
  ${BackgroundImageCommon}
  display: ${(props) => (props.display ? props.display : "inline-block")};

  width: ${(props) => props.width};
  height: ${(props) => props.height};

  background-image: url(${(props) => props.url});
  background-size: ${(props) => `${props.width} ${props.height}`};
  background-position: ${(props) =>
    props.bgPosition ? props.bgPosition : "center"};
`;

export const BackgroundImageA = styled.a<{
  url: string;
  width: string;
  height: string;
  display?: CSSProperties["display"];
  bgPosition?: CSSProperties["backgroundPosition"];
}>`
  ${BackgroundImageCommon}
  display: ${(props) => (props.display ? props.display : "inline-block")};

  width: ${(props) => props.width};
  height: ${(props) => props.height};

  background-image: url(${(props) => props.url});
  background-size: ${(props) => `${props.width} ${props.height}`};
  background-position: ${(props) =>
    props.bgPosition ? props.bgPosition : "center"};
`;

export const BackgroundImageDiv = styled.div<{
  url: string;
  width: string;
  height: string;
  display?: CSSProperties["display"];

  bgPosition?: CSSProperties["backgroundPosition"];
}>`
  ${BackgroundImageCommon}
  display: ${(props) => (props.display ? props.display : "block")};

  width: ${(props) => props.width};
  height: ${(props) => props.height};

  background-image: url(${(props) => props.url});
  background-size: ${(props) => `${props.width} ${props.height}`};
  background-position: ${(props) =>
    props.bgPosition ? props.bgPosition : "center"};
`;
