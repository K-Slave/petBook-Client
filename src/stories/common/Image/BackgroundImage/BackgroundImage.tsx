import React, { AnchorHTMLAttributes } from "react";
import styled, { CSSProperties, css } from "styled-components";
import { BackgroundImageA, BackgroundImageSpan } from "./BackgroundImage.style";

const BgImgA = styled(BackgroundImageA)``;
const BgImgSpan = styled(BackgroundImageSpan)``;

interface Props {
  url: string;
  width: string;
  height: string;
  tag?: "span" | "a" | "div";
  anchorAttr?: AnchorHTMLAttributes<HTMLAnchorElement>;
  display?: CSSProperties["display"];
  bgPosition?: CSSProperties["backgroundPosition"];
}

// Styled Components를 사용한 예시
// Overriding 하여 사용하기 권장
const BackgroundImage = ({
  url,
  width,
  height,
  display,
  bgPosition,
  tag,
  anchorAttr,
}: Props) => {
  if (tag === "div") {
    return (
      <BgImgSpan
        url={url}
        width={width}
        height={height}
        display={display}
        bgPosition={bgPosition}
      />
    );
  }

  if (tag === "a") {
    return (
      <BgImgA
        url={url}
        width={width}
        height={height}
        display={display}
        bgPosition={bgPosition}
        {...anchorAttr}
      />
    );
  }

  return (
    <BgImgSpan
      url={url}
      width={width}
      height={height}
      display={display}
      bgPosition={bgPosition}
    />
  );
};

export default BackgroundImage;
