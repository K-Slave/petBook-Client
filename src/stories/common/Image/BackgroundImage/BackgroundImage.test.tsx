import React from "react";
import BackgroundImage from "./BackgroundImage";
import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";

const originUrl =
  process.env.NODE_ENV === "production"
    ? "https://petbook.site"
    : "http://localhost:3000";

const TestComponent = () => {
  return (
    <BackgroundImage
      url={`${originUrl}/icons/icons8-backjo-96.png`}
      width="3.125rem"
      height="3.125rem"
      bgPosition="center"
      tag="span"
    />
  );
};

describe("BackgroundImage", () => {
  it("should render without errors", () => {
    const { container } = render(<TestComponent />);

    expect(container).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const tree = renderer.create(<TestComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders the correct content", () => {
    render(<TestComponent />);

    // 정상적으로 렌더링되었는지 확인
    const spanElement = document.querySelector("span");

    expect(spanElement).toBeInTheDocument();

    // Check the width attribute of the span element
    expect(spanElement).toHaveStyle({ width: "3.125rem" });
  });

  test('renders a div tag when tag prop is "div"', () => {
    render(
      <BackgroundImage
        url={`${originUrl}/Landing__Background.png`}
        width="100px"
        height="100px"
        tag="div"
      />
    );
    expect(document.querySelector("div")).toBeInTheDocument();
  });

  test('renders an a tag when tag prop is "a"', () => {
    render(
      <BackgroundImage
        url={`${originUrl}/icons/icons8-gosum-96.png`}
        width="100px"
        height="100px"
        tag="a"
        anchorAttr={{
          href: "https://icons8.com/icon/aKEeT9FEK2cQ/%EA%B3%A0%EC%8A%B4%EB%8F%84%EC%B9%98",
        }}
      />
    );
    const linkElement = screen.getByRole("link");
    expect(linkElement).toBeInTheDocument();
    expect(linkElement).toHaveAttribute(
      "href",
      "https://icons8.com/icon/aKEeT9FEK2cQ/%EA%B3%A0%EC%8A%B4%EB%8F%84%EC%B9%98"
    );
  });
});
