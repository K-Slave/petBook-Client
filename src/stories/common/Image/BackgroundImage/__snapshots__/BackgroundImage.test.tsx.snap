// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundImage matches snapshot 1`] = `
<span
  className="sc-bczRLJ sc-jSMfEi hzVEgG"
  height="3.125rem"
  width="3.125rem"
/>
`;
